airbyte-integrations/connectors/source-google-drive/source_google_drive/stream_reader.py

```diff
@@ -0,0 +1,195 @@
+                self._drive_service = build("drive", "v3", credentials=creds)
+        except Exception as e:
+            raise AirbyteTracedException(
+                internal_message=str(e),
+                message="Could not authenticate with Google Drive. Please check your credentials.",
+                failure_type=FailureType.config_error,
+                exception=e,
+            )
+
+        return self._drive_service
+
+    def get_matching_files(self, globs: List[str], prefix: Optional[str], logger: logging.Logger) -> Iterable[RemoteFile]:
+        """
+        Get all files matching the specified glob patterns.
+        """
+        service = self.google_drive_service
+        root_folder_id = self._get_folder_id(self.config.folder_url)
+        # ignore prefix argument as it's legacy only and this is a new connector
+        prefixes = self.get_prefixes_from_globs(globs)
+
+        folder_id_queue = [("", root_folder_id)]
+        seen: Set[str] = set()
+        while len(folder_id_queue) > 0:
+            (path, folder_id) = folder_id_queue.pop()
+            # fetch all files in this folder (1000 is the max page size)
+            request = service.files().list(
+                q=f"'{folder_id}' in parents", pageSize=1000, fields="nextPageToken, files(id, name, modifiedTime, mimeType)"
+            )
+            while True:
+                results = request.execute()
+                new_files = results.get("files", [])
+                for new_file in new_files:
+                    # It's possible files and folders are linked up multiple times, this prevents us from getting stuck in a loop
+                    if new_file["id"] in seen:
+                        continue
+                    seen.add(new_file["id"])
+                    file_name = path + self._get_file_name(new_file)
+                    if new_file["mimeType"] == FOLDER_MIME_TYPE:
+                        folder_name = f"{file_name}/"
+                        # check prefix matching in both directions to handle
+                        prefix_matches_folder_name = any(prefix.startswith(folder_name) for prefix in prefixes)
+                        folder_name_matches_prefix = any(folder_name.startswith(prefix) for prefix in prefixes)
+                        if prefix_matches_folder_name or folder_name_matches_prefix or len(prefixes) == 0:
+                            folder_id_queue.append((folder_name, new_file["id"]))
+                        continue
+                    else:
+                        last_modified = datetime.strptime(new_file["modifiedTime"], "%Y-%m-%dT%H:%M:%S.%fZ")
+                        remote_file = GoogleDriveRemoteFile(
+                            uri=file_name, last_modified=last_modified, id=new_file["id"], mimeType=new_file["mimeType"]
+                        )
+                        if self.file_matches_globs(remote_file, globs):
+                            yield remote_file
```

Comment:
Following also my comments in `_get_file_name()`, it doesn't feel right to use our appended suffixes (.docx, .pdf) for documents that don't have those suffixes. Could we remove that suffix without breaking any key functionality?

Also, just to confirm, do we process _all_ files in the included directory, except those excluded by a users' glob pattern? Or do we have some kind of internal filter when using the document parser? I'm a bit worried this could break if a non-readable non-text file (.zip, .ico, .parquet) exists in the directory.

Reply:
As mentioned in the other comment, once https://github.com/airbytehq/airbyte/pull/31997 is merged we shouldn't need to append suffixes anymore.

> Also, just to confirm, do we process all files in the included directory, except those excluded by a users' glob pattern? Or do we have some kind of internal filter when using the document parser? I'm a bit worried this could break if a non-readable non-text file (.zip, .ico, .parquet) exists in the directory.

As the fact whether a file is readable or not is a property of the parser, not the stream reader, this can't be controlled here. See our previous discussion of this over here: https://github.com/airbytehq/airbyte/pull/31209#discussion_r1355760696

Reply:
I like the idea of a failure with warnings. That seems like the best of both worlds to me.